pkg/streamingpromql/optimize/ast/propagate_matchers.go

```diff
@@ -0,0 +1,312 @@
+		return nil, nil
+	}
+}
+
+func (mapper *propagateMatchers) extractVectorSelectorsFromAggregateExpr(e *parser.AggregateExpr) ([]*enrichedVectorSelector, []*labels.Matcher) {
+	include := !e.Without
+	if len(e.Grouping) == 0 && include {
+		// Shortcut if there are no labels allowed to propagate inwards or outwards.
+		return nil, nil
+	}
+	vss, labelMatchers := mapper.extractVectorSelectors(e.Expr)
+	if len(vss) == 0 {
+		return nil, nil
+	}
+	groupingSet := newStringSet(e.Grouping)
+	for _, vs := range vss {
+		updateVectorSelectorWithAggregationInfo(vs, groupingSet, include)
+	}
+	newMatchers := mapper.getMatchersToPropagate(labelMatchers, groupingSet, include)
+	return vss, newMatchers
+}
+
+func updateVectorSelectorWithAggregationInfo(vs *enrichedVectorSelector, groupingSet stringSet, include bool) {
+	if vs.containsAgg {
+		switch {
+		case vs.include && include:
+			vs.labelsSet = vs.labelsSet.Intersection(groupingSet)
+		case !vs.include && !include:
+			vs.labelsSet = vs.labelsSet.Union(groupingSet)
+		case vs.include && !include:
+			vs.labelsSet = vs.labelsSet.Difference(groupingSet)
+		case !vs.include && include:
+			vs.labelsSet = groupingSet.Difference(vs.labelsSet)
+			vs.include = true
+		}
+		return
+	}
+	vs.labelsSet = groupingSet
+	vs.include = include
+	vs.containsAgg = true
+}
+
+func VectorSelectorArgumentIndex(funcName string) (int, error) {
+	// If the function is eligible for matcher propagation, return the index of the argument
+	// that contains the vector selector (may be embedded in other kinds of expressions).
+	// Otherwise, return -1.
+	switch funcName {
+	// Simple
+	case "absent", "changes", "resets":
+		return 0, nil
+	// Mathematical
+	case "abs", "sgn", "sqrt", "exp", "deriv", "ln", "log2", "log10", "ceil", "floor", "round", "clamp", "clamp_min", "clamp_max":
+		return 0, nil
+	// More complicated mathematical functions
+	case "double_exponential_smoothing", "predict_linear":
+		return 0, nil
+	// Trigonometric
+	case "acos", "acosh", "asin", "asinh", "atan", "atanh", "cos", "cosh", "sin", "sinh", "tan", "tanh", "deg", "rad":
+		return 0, nil
+	// Differences between samples (using matrix selectors)
+	case "rate", "delta", "increase", "idelta", "irate":
+		return 0, nil
+	// Histogram-related
+	case "histogram_avg", "histogram_count", "histogram_sum", "histogram_stddev", "histogram_stdvar":
+		return 0, nil
+	case "histogram_quantile":
+		return 1, nil
+	case "histogram_fraction":
+		return 2, nil
+	// Aggregation over time
+	case "avg_over_time", "min_over_time", "max_over_time", "sum_over_time", "count_over_time", "stddev_over_time", "stdvar_over_time", "last_over_time", "present_over_time", "absent_over_time", "mad_over_time", "ts_of_min_over_time", "ts_of_max_over_time", "ts_of_last_over_time":
+		return 0, nil
+	case "quantile_over_time":
+		return 1, nil
+	// Time
+	case "minute", "hour", "day_of_week", "day_of_month", "day_of_year", "days_in_month", "month", "year", "timestamp":
+		return 0, nil
```

Comment:
For these functions, the argument is optional - we might need to add a check in `extractVectorSelectors` to make sure we don't try to read an argument that's not there (and some tests to cover these cases 🙂)

Reply:
Good catch, thanks!